File: visualization/src/types/workspace.ts
/** Workspace manifest, resource and CAS contracts. */

export interface WorkspaceResourceRecord {
  link: string;
  relative_path: string;
  kind: string;
  media_type: string;
  suffix?: string;
  summary: string;
  size: number;
  mtime_ns: number;
  digest?: string;
  description?: string;
  described_digest?: string;
  retention: "ephemeral" | "turn" | "day" | "persistent";
  owner_turn_id?: string;
}

export interface WorkspaceManifest {
  schema_version: number;
  day: string;
  revision: number;
  resources: WorkspaceResourceRecord[];
}

export interface WorkspaceTextRead {
  link: string;
  text: string;
  truncated: boolean;
  size: number;
  digest: string;
}

export interface WorkspaceBlobHeaders {
  link: string;
  digest: string;
  size: number;
}

export interface WorkspaceWriteResponse {
  record: WorkspaceResourceRecord;
  manifest: WorkspaceManifest;
}

export interface TrashItem {
  ref: string;
  link: string;
  relative_path: string;
  kind: string;
  media_type: string;
  size: number;
  digest?: string;
  moved_at: number;
}

export interface WorkspaceWriteRequest {
  link: string;
  text: string;
  overwrite: boolean;
  expected_digest: string;
  expected_revision: number;
  retention?: WorkspaceResourceRecord["retention"];
}

export interface WorkspaceTrashRequest {
  link: string;
  expected_digest: string;
  expected_revision: number;
}

export interface WorkspaceRestoreRequest {
  trash_ref: string;
  expected_revision: number;
}

File: tinysoul/endpoint/http/schemas/workspace.py
"""Workspace read and write request schemas."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class WorkspaceWriteRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    link: str = Field(min_length=1)
    text: str
    overwrite: bool = False
    expected_digest: str = ""
    expected_revision: int = Field(ge=0)
    retention: Literal["ephemeral", "turn", "day", "persistent"] | None = None


class WorkspaceTrashRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    link: str = Field(min_length=1)
    expected_digest: str = Field(min_length=1)
    expected_revision: int = Field(ge=0)


class WorkspaceRestoreRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    trash_ref: str = Field(min_length=1)
    expected_revision: int = Field(ge=0)
